crates/sui-adapter/src/programmable_transactions/execution.rs

```diff
@@ -436,6 +447,92 @@ fn execute_move_publish<E: fmt::Debug, S: StorageView<E>, Mode: ExecutionMode>(
     Ok(values)
 }
 
+/// Upgrade a Move package.  Returns an `UpgradeReceipt` for the upgraded package on success.
+fn execute_move_upgrade<E: fmt::Debug, S: StorageView<E>>(
+    context: &mut ExecutionContext<E, S>,
+    upgrade_ticket_arg: Argument,
+    dep_ids: Vec<ObjectID>,
+    module_bytes: Vec<Vec<u8>>,
+) -> Result<Value, ExecutionError> {
+    // Check that package upgrades are supported.
+    context
+        .protocol_config
+        .check_package_upgrades_supported()
+        .map_err(|_| {
+            ExecutionError::from_kind(ExecutionErrorKind::PackageUpgradeError(
+                PackageUpgradeError::NotSupported,
+            ))
+        })?;
+
+    assert_invariant!(
+        !module_bytes.is_empty(),
+        "empty package is checked in transaction input checker"
+    );
+
+    let upgrade_ticket: UpgradeTicket = {
+        let mut ticket_bytes = Vec::new();
+        let ticket_obj_val: ObjectValue =
+            context.by_value_arg(CommandKind::Upgrade, 0, upgrade_ticket_arg)?;
+        ticket_obj_val.write_bcs_bytes(&mut ticket_bytes);
```

Comment:
You need to type check the `ticket_obj_val` before doing `bcs::from_bytes` or you are going to accidentally support type conversion for upgrade tickets... which would be bad :)